# Before your start:
- Read the README.md file
- Comment as much as you can and use the resources (README.md file)
- Happy learning!

In [25]:
# import numpy and pandas
import numpy as np
import pandas as pd
import scipy
from scipy import stats

# Challenge 1 - Exploring the Data

In this challenge, we will examine all salaries of employees of the City of Chicago. We will start by loading the dataset and examining its contents

In [4]:
# Run this code:

salaries = pd.read_csv('../Current_Employee_Names__Salaries__and_Position_Titles.csv')

Examine the `salaries` dataset using the `head` function below.

In [5]:
# Your code here:
salaries.head()

Name                              Job Titles  \
0    AARON,  JEFFERY M                                SERGEANT   
1      AARON,  KARINA   POLICE OFFICER (ASSIGNED AS DETECTIVE)   
2  AARON,  KIMBERLEI R                CHIEF CONTRACT EXPEDITER   
3  ABAD JR,  VICENTE M                       CIVIL ENGINEER IV   
4    ABASCAL,  REECE E             TRAFFIC CONTROL AIDE-HOURLY   

         Department Full or Part-Time Salary or Hourly  Typical Hours  \
0            POLICE                 F           Salary            NaN   
1            POLICE                 F           Salary            NaN   
2  GENERAL SERVICES                 F           Salary            NaN   
3       WATER MGMNT                 F           Salary            NaN   
4              OEMC                 P           Hourly           20.0   

   Annual Salary  Hourly Rate  
0       101442.0          NaN  
1        94122.0          NaN  
2       101592.0          NaN  
3       110064.0          NaN  
4            NaN        19.86

We see from looking at the `head` function that there is quite a bit of missing data. Let's examine how much missing data is in each column. Produce this output in the cell below

In [8]:
# Your code here:
salaries['Typical Hours'].isnull().sum()

25161

In [9]:
salaries.isnull().sum()

Name                     0
Job Titles               0
Department               0
Full or Part-Time        0
Salary or Hourly         0
Typical Hours        25161
Annual Salary         8022
Hourly Rate          25161
dtype: int64

Let's also look at the count of hourly vs. salaried employees. Write the code in the cell below

In [16]:
# Your code here:
salaries.groupby('Salary or Hourly').count()['Name']

Salary or Hourly
Hourly     8022
Salary    25161
Name: Name, dtype: int64

What this information indicates is that the table contains information about two types of employees - salaried and hourly. Some columns apply only to one type of employee while other columns only apply to another kind. This is why there are so many missing values. Therefore, we will not do anything to handle the missing values.

There are different departments in the city. List all departments and the count of employees in each department.

In [19]:
# Your code here:
salaries.groupby('Department').count()['Name']

Department
ADMIN HEARNG                39
ANIMAL CONTRL               81
AVIATION                  1629
BOARD OF ELECTION          107
BOARD OF ETHICS              8
BUDGET & MGMT               46
BUILDINGS                  269
BUSINESS AFFAIRS           171
CITY CLERK                  84
CITY COUNCIL               411
COMMUNITY DEVELOPMENT      207
COPA                       116
CULTURAL AFFAIRS            65
DISABILITIES                28
DoIT                        99
FAMILY & SUPPORT           615
FINANCE                    560
FIRE                      4641
GENERAL SERVICES           980
HEALTH                     488
HUMAN RELATIONS             16
HUMAN RESOURCES             79
INSPECTOR GEN               87
LAW                        407
LICENSE APPL COMM            1
MAYOR'S OFFICE              85
OEMC                      2102
POLICE                   13414
POLICE BOARD                 2
PROCUREMENT                 92
PUBLIC LIBRARY            1015
STREETS & SAN             21

# Challenge 2 - Hypothesis Tests

In this section of the lab, we will test whether the hourly wage of all hourly workers is significantly different from $30/hr. Import the correct one sample test function from scipy and perform the hypothesis test for a 95% two sided confidence interval.

In [22]:
salaries['Hourly Rate'].describe()

count    8022.000000
mean       32.788558
std        12.112573
min         2.650000
25%        21.200000
50%        35.600000
75%        40.200000
max       109.000000
Name: Hourly Rate, dtype: float64

In [21]:
table= pd.pivot_table(salaries, values='Hourly Rate', index=['Salary or Hourly'])
table

Hourly Rate
Salary or Hourly             
Hourly              32.788558

In [23]:
# Your code here:

Z=(32.79-30)/(12.112573/(8022**.5))
Z

20.63047011941587

In [45]:
hourlyonly=salaries.dropna(subset=['Hourly Rate'])
hourlyonly

Name                      Job Titles  \
4           ABASCAL,  REECE E     TRAFFIC CONTROL AIDE-HOURLY   
6       ABBATACOLA,  ROBERT J             ELECTRICAL MECHANIC   
7          ABBATE,  JOSEPH L          POOL MOTOR TRUCK DRIVER   
10           ABBOTT,  BETTY L              FOSTER GRANDPARENT   
18       ABDULLAH,  LAKENYA N                  CROSSING GUARD   
...                       ...                             ...   
33164        ZUREK,  FRANCIS              ELECTRICAL MECHANIC   
33168  ZWARYCZ MANN,  IRENE A                  CROSSING GUARD   
33169      ZWARYCZ,  THOMAS J         POOL MOTOR TRUCK DRIVER   
33174        ZYGADLO,  JOHN P          MACHINIST (AUTOMOTIVE)   
33175     ZYGADLO,  MICHAEL J  FRM OF MACHINISTS - AUTOMOTIVE   

             Department Full or Part-Time Salary or Hourly  Typical Hours  \
4                  OEMC                 P           Hourly           20.0   
6              AVIATION                 F           Hourly           40.0   
7         STREETS & SAN                 F           Hourly           40.0   
10     FAMILY & SUPPORT                 P           Hourly           20.0   
18                 OEMC                 P           Hourly           20.0   
...                 ...               ...              ...            ...   
33164              OEMC                 F           Hourly           40.0   
33168              OEMC                 P           Hourly           20.0   
33169       WATER MGMNT                 F           Hourly           40.0   
33174  GENERAL SERVICES                 F           Hourly           40.0   
33175  GENERAL SERVICES                 F           Hourly           40.0   

       Annual Salary  Hourly Rate  svh  
4                NaN        19.86    0  
6                NaN        46.10    0  
7                NaN        35.60    0  
10               NaN         2.65    0  
18               NaN        17.68    0  
...              ...          ...  ...  
33164            NaN        46.10    0  
33168            NaN        17.68    0  
33169            NaN        35.60    0  
33174            NaN        46.35    0  
33175            NaN        48.85    0  

[8022 rows x 9 columns]

In [36]:
from statsmodels.stats.weightstats import ztest as ztest

In [46]:
ztest(hourlyonly['Hourly Rate'], value=30)

(20.6198057854942, 1.8228873859286195e-94)

**Yes. Critical value corresponding to an alpha of .05 is 1.6449 and our Zscore is 20.63 hence we can confidently say that the hourly wage of all hourly workers is significantly different from $30/hr.**

We are also curious about salaries in the police force. The chief of police in Chicago claimed in a press briefing that salaries this year are higher than last year's mean of $86000/year a year for all salaried employees. Test this one sided hypothesis using a 95% confidence interval.

Hint: A one tailed test has a p-value that is half of the two tailed p-value. If our hypothesis is greater than, then to reject, the test statistic must also be positive.

In [49]:
# Your code here:
police_dept=salaries[(salaries['Department']=='POLICE')]
police_dept.describe()

Typical Hours  Annual Salary  Hourly Rate           svh
count           10.0   13404.000000    10.000000  13414.000000
mean            20.0   86486.414503     9.384000      0.999255
std              0.0   18272.228294     0.098116      0.027294
min             20.0   38376.000000     9.270000      0.000000
25%             20.0   84054.000000     9.270000      1.000000
50%             20.0   90024.000000     9.460000      1.000000
75%             20.0   94524.000000     9.460000      1.000000
max             20.0  260004.000000     9.460000      1.000000

In [55]:
police_salaries=police_dept.dropna(subset=['Annual Salary'])
police_salaries

Name                              Job Titles  \
0           AARON,  JEFFERY M                                SERGEANT   
1             AARON,  KARINA   POLICE OFFICER (ASSIGNED AS DETECTIVE)   
9            ABBATE,  TERRY M                          POLICE OFFICER   
11           ABDALLAH,  ZAID                           POLICE OFFICER   
12     ABDELHADI,  ABDALMAHD                           POLICE OFFICER   
...                       ...                                     ...   
33177        ZYGMUNT,  DAWID                           POLICE OFFICER   
33178   ZYLINSKA,  KATARZYNA                           POLICE OFFICER   
33179      ZYMANTAS,  LAURA C                          POLICE OFFICER   
33180       ZYMANTAS,  MARK E                          POLICE OFFICER   
33181     ZYRKOWSKI,  CARLO E                          POLICE OFFICER   

      Department Full or Part-Time Salary or Hourly  Typical Hours  \
0         POLICE                 F           Salary            NaN   
1         POLICE                 F           Salary            NaN   
9         POLICE                 F           Salary            NaN   
11        POLICE                 F           Salary            NaN   
12        POLICE                 F           Salary            NaN   
...          ...               ...              ...            ...   
33177     POLICE                 F           Salary            NaN   
33178     POLICE                 F           Salary            NaN   
33179     POLICE                 F           Salary            NaN   
33180     POLICE                 F           Salary            NaN   
33181     POLICE                 F           Salary            NaN   

       Annual Salary  Hourly Rate  svh  
0           101442.0          NaN    1  
1            94122.0          NaN    1  
9            93354.0          NaN    1  
11           84054.0          NaN    1  
12           87006.0          NaN    1  
...              ...          ...  ...  
33177        72510.0          NaN    1  
33178        72510.0          NaN    1  
33179        48078.0          NaN    1  
33180        90024.0          NaN    1  
33181        93354.0          NaN    1  

[13404 rows x 9 columns]

In [130]:
z_value,p_value=ztest(police_salaries['Annual Salary'], value=86000)
one_tail=p_value/2
one_tail

0.0010280845287547855

In [57]:
Z2=(86486.414503-86000)/(18272.228294/(13404**.5))
Z2

3.0819970048666927

**As we can see, the p-value is well below 0.05. So we reject H0, meaning the new mean is statistically significantly different.** 

Using the `crosstab` function, find the department that has the most hourly workers. 

In [67]:
# Your code here:
pd.crosstab(index=salaries['Department'], columns=salaries['Salary or Hourly']).sort_values(by='Hourly', ascending=False)[:1]

Salary or Hourly  Hourly  Salary
Department                      
STREETS & SAN       1862     336

The workers from the department with the most hourly workers have complained that their hourly wage is less than $35/hour. Using a one sample t-test, test this one-sided hypothesis at the 95% confidence level.

In [74]:
# Your code here:
Hourly=salaries[(salaries['Salary or Hourly']=='Hourly')]
SNSN=salaries[(salaries['Department']=='STREETS & SAN')]
SNS=SNSN.dropna(subset=['Hourly Rate'])
SNS

Name               Job Titles     Department  \
7          ABBATE,  JOSEPH L   POOL MOTOR TRUCK DRIVER  STREETS & SAN   
21      ABDUL-SHAKUR,  TAHIR     GENERAL LABORER - DSS  STREETS & SAN   
24     ABERCROMBIE,  TIMOTHY        MOTOR TRUCK DRIVER  STREETS & SAN   
36        ABRAMS,  DANIELLE T       SANITATION LABORER  STREETS & SAN   
39          ABRAMS,  SAMUEL A  POOL MOTOR TRUCK DRIVER  STREETS & SAN   
...                       ...                      ...            ...   
33106        ZIZUMBO,  JOSE N       MOTOR TRUCK DRIVER  STREETS & SAN   
33107         ZIZUMBO,  LUIS        MOTOR TRUCK DRIVER  STREETS & SAN   
33147        ZUMMO,  ROBERT J       MOTOR TRUCK DRIVER  STREETS & SAN   
33149     ZUNICH,  JONATHAN G       SANITATION LABORER  STREETS & SAN   
33156        ZUNIGA,  THOMAS     GENERAL LABORER - DSS  STREETS & SAN   

      Full or Part-Time Salary or Hourly  Typical Hours  Annual Salary  \
7                     F           Hourly           40.0            NaN   
21                    F           Hourly           40.0            NaN   
24                    F           Hourly           40.0            NaN   
36                    F           Hourly           40.0            NaN   
39                    F           Hourly           40.0            NaN   
...                 ...              ...            ...            ...   
33106                 F           Hourly           40.0            NaN   
33107                 F           Hourly           40.0            NaN   
33147                 F           Hourly           40.0            NaN   
33149                 F           Hourly           40.0            NaN   
33156                 F           Hourly           40.0            NaN   

       Hourly Rate  svh  
7            35.60    0  
21           21.43    0  
24           35.60    0  
36           36.21    0  
39           35.60    0  
...            ...  ...  
33106        36.13    0  
33107        35.60    0  
33147        35.60    0  
33149        36.21    0  
33156        22.12    0  

[1862 rows x 9 columns]

In [121]:
stats.ttest_1samp(SNS['Hourly Rate'], popmean=35, alternative='less')

Ttest_1sampResult(statistic=-9.567447887848152, pvalue=1.6689265282353859e-21)

**They are correct.**

# Challenge 3 - Constructing Confidence Intervals

While testing our hypothesis is a great way to gather empirical evidence for accepting or rejecting the hypothesis, another way to gather evidence is by creating a confidence interval. A confidence interval gives us information about the true mean of the population. So for a 95% confidence interval, we are 95% sure that the mean of the population is within the confidence interval. 
).

To read more about confidence intervals, click [here](https://en.wikipedia.org/wiki/Confidence_interval).


In the cell below, we will construct a 95% confidence interval for the mean hourly wage of all hourly workers. 

The confidence interval is computed in SciPy using the `t.interval` function. You can read more about this function [here](https://docs.scipy.org/doc/scipy-0.14.0/reference/generated/scipy.stats.t.html).

To compute the confidence interval of the hourly wage, use the 0.95 for the confidence level, number of rows - 1 for degrees of freedom, the mean of the sample for the location parameter and the standard error for the scale. The standard error can be computed using [this](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.sem.html) function in SciPy.

In [93]:
# Your code here:
Hourly.describe()

Typical Hours  Annual Salary  Hourly Rate     svh
count    8022.000000            0.0  8022.000000  8022.0
mean       34.507604            NaN    32.788558     0.0
std         9.252077            NaN    12.112573     0.0
min        10.000000            NaN     2.650000     0.0
25%        20.000000            NaN    21.200000     0.0
50%        40.000000            NaN    35.600000     0.0
75%        40.000000            NaN    40.200000     0.0
max        40.000000            NaN   109.000000     0.0

In [108]:
error=stats.sem(Hourly['Hourly Rate'], axis=0, ddof=1)
error

0.1352368565101596

In [103]:
from scipy.stats import t

In [107]:
t.interval(alpha=0.95, df=(len(Hourly))-1, loc=32.788558, scale=error)

(32.52345862860401, 33.05365737139599)

Now construct the 95% confidence interval for all salaried employeed in the police in the cell below.

In [125]:
# Your code here:
salpol=salaries[(salaries['Salary or Hourly']=='Salary')&(salaries['Department']=='POLICE')]
salpol.describe()

Typical Hours  Annual Salary  Hourly Rate      svh
count            0.0   13404.000000          0.0  13404.0
mean             NaN   86486.414503          NaN      1.0
std              NaN   18272.228294          NaN      0.0
min              NaN   38376.000000          NaN      1.0
25%              NaN   84054.000000          NaN      1.0
50%              NaN   90024.000000          NaN      1.0
75%              NaN   94524.000000          NaN      1.0
max              NaN  260004.000000          NaN      1.0

In [126]:
errorsal=stats.sem(salpol['Annual Salary'], axis=0, ddof=1)
errorsal

157.82445675052244

In [120]:
t.interval(alpha=0.95, df=(len(sal))-1, loc=86486.414503, scale=errorsal)

(86177.06937032643, 86795.75963567359)

# Challenge 4 - Hypothesis Tests of Proportions

Another type of one sample test is a hypothesis test of proportions. In this test, we examine whether the proportion of a group in our sample is significantly different than a fraction. 

You can read more about one sample proportion tests [here](http://sphweb.bumc.bu.edu/otlt/MPH-Modules/BS/SAS/SAS6-CategoricalData/SAS6-CategoricalData2.html).

In the cell below, use the `proportions_ztest` function from `statsmodels` to perform a hypothesis test that will determine whether the number of hourly workers in the City of Chicago is significantly different from 25% at the 95% confidence level.

In [79]:
Hourly=salaries[(salaries['Salary or Hourly']=='Hourly')]
Hourly.head(3)

Name                   Job Titles     Department  \
4      ABASCAL,  REECE E  TRAFFIC CONTROL AIDE-HOURLY           OEMC   
6  ABBATACOLA,  ROBERT J          ELECTRICAL MECHANIC       AVIATION   
7     ABBATE,  JOSEPH L       POOL MOTOR TRUCK DRIVER  STREETS & SAN   

  Full or Part-Time Salary or Hourly  Typical Hours  Annual Salary  \
4                 P           Hourly           20.0            NaN   
6                 F           Hourly           40.0            NaN   
7                 F           Hourly           40.0            NaN   

   Hourly Rate  svh  
4        19.86    0  
6        46.10    0  
7        35.60    0

In [80]:
Hourly.describe()

Typical Hours  Annual Salary  Hourly Rate     svh
count    8022.000000            0.0  8022.000000  8022.0
mean       34.507604            NaN    32.788558     0.0
std         9.252077            NaN    12.112573     0.0
min        10.000000            NaN     2.650000     0.0
25%        20.000000            NaN    21.200000     0.0
50%        40.000000            NaN    35.600000     0.0
75%        40.000000            NaN    40.200000     0.0
max        40.000000            NaN   109.000000     0.0

In [83]:
# Your code here:
from statsmodels.stats.proportion import proportions_ztest
proportions_ztest(count=len(Hourly['Hourly Rate']), nobs=len(salaries), value=0.25)


(-3.5099964213703005, 0.0004481127249057967)

**Yes it's different**